### PR TITLE
Add functionality specific to Giantnetes.

### PR DESCRIPTION
**Pull request by r7vme**
1) Get guest cluster access from cert-operator secret
2) Switch to microerror and micrologger

All PRs (all tested together):
https://github.com/giantswarm/kvm-operator-node-controller/pull/4
https://github.com/giantswarm/kubernetesd/pull/289
https://github.com/giantswarm/kvm-operator/pull/221
https://github.com/giantswarm/kvmtpr/pull/51

### CODE REVIEW

**Comment by r7vme** [comment]
Here is how logs look like now. Right after cluster start up: 1) API is not available 2) all nodes still NotReady, but pods present so skip 3) Regular routine when all nodes are Ready

```
{"Info":"Trying to use in-cluster config","caller":"github.com/giantswarm/kvm-operator-node-controller/provider/cloud.go:61","time":"17-10-26 19:51:36.517"}
{"Info":"Starting kvm-operator node controller","caller":"kvm-operator-node-controller/controller.go:156","time":"17-10-26 19:51:36.518"}
{"Error":"Error monitoring node status","Trace":"Get https://api.8842v.g8s.fra-1.giantswarm.io/api/v1/nodes?resourceVersion=0: EOF","caller":"kvm-operator-node-controller/controller.go:178","time
":"17-10-26 19:51:36.561"}
{"Error":"Error monitoring node status","Trace":"Get https://api.8842v.g8s.fra-1.giantswarm.io/api/v1/nodes?resourceVersion=0: EOF","caller":"kvm-operator-node-controller/controller.go:178","time
":"17-10-26 19:52:06.622"}
{"Error":"Error monitoring node status","Trace":"Get https://api.8842v.g8s.fra-1.giantswarm.io/api/v1/nodes?resourceVersion=0: EOF","caller":"kvm-operator-node-controller/controller.go:178","time
":"17-10-26 19:52:36.635"}
{"Error":"Error monitoring node status","Trace":"Get https://api.8842v.g8s.fra-1.giantswarm.io/api/v1/nodes?resourceVersion=0: EOF","caller":"kvm-operator-node-controller/controller.go:178","time
":"17-10-26 19:53:06.648"}
{"Error":"Error monitoring node status","Trace":"Get https://api.8842v.g8s.fra-1.giantswarm.io/api/v1/nodes?resourceVersion=0: EOF","caller":"kvm-operator-node-controller/controller.go:178","time
":"17-10-26 19:53:36.653"}
{"Info":"Checking node status","caller":"kvm-operator-node-controller/controller.go:186","node":"worker-k9nl5-7148567-prxmm","time":"17-10-26 19:54:06.675"}
{"Info":"Checking pod","caller":"github.com/giantswarm/kvm-operator-node-controller/provider/instances.go:60","namespace":"8842v","pod":"worker-k9nl5-7148567-prxmm","time":"17-10-26 19:54:06.675"
}
{"Info":"Pod is ready","caller":"github.com/giantswarm/kvm-operator-node-controller/provider/instances.go:78","namespace":"8842v","pod":"worker-k9nl5-7148567-prxmm","time":"17-10-26 19:54:06.720"
}
{"Info":"Checking node status","caller":"kvm-operator-node-controller/controller.go:186","node":"worker-us4tg-1197729831-lp2tw","time":"17-10-26 19:54:06.720"}
{"Info":"Checking pod","caller":"github.com/giantswarm/kvm-operator-node-controller/provider/instances.go:60","namespace":"8842v","pod":"worker-us4tg-1197729831-lp2tw","time":"17-10-26 19:54:06.7
20"}
{"Info":"Pod is ready","caller":"github.com/giantswarm/kvm-operator-node-controller/provider/instances.go:78","namespace":"8842v","pod":"worker-us4tg-1197729831-lp2tw","time":"17-10-26 19:54:06.7
24"}
{"Info":"Checking node status","caller":"kvm-operator-node-controller/controller.go:186","node":"worker-v3v91-1995990299-s6nkx","time":"17-10-26 19:54:06.724"}
{"Info":"Checking pod","caller":"github.com/giantswarm/kvm-operator-node-controller/provider/instances.go:60","namespace":"8842v","pod":"worker-v3v91-1995990299-s6nkx","time":"17-10-26 19:54:06.7
24"}
{"Info":"Pod is ready","caller":"github.com/giantswarm/kvm-operator-node-controller/provider/instances.go:78","namespace":"8842v","pod":"worker-v3v91-1995990299-s6nkx","time":"17-10-26 19:54:06.7
27"}
{"Info":"Checking node status","caller":"kvm-operator-node-controller/controller.go:186","node":"master-2ir56-3922763829-nq9wf","time":"17-10-26 19:54:06.727"}
{"Info":"Checking pod","caller":"github.com/giantswarm/kvm-operator-node-controller/provider/instances.go:60","namespace":"8842v","pod":"master-2ir56-3922763829-nq9wf","time":"17-10-26 19:54:06.7
27"}
{"Info":"Pod is ready","caller":"github.com/giantswarm/kvm-operator-node-controller/provider/instances.go:78","namespace":"8842v","pod":"master-2ir56-3922763829-nq9wf","time":"17-10-26 19:54:06.7
30"}
{"Info":"Checking node status","caller":"kvm-operator-node-controller/controller.go:186","node":"worker-us4tg-1197729831-lp2tw","time":"17-10-26 19:54:36.733"}
{"Info":"node state","caller":"kvm-operator-node-controller/controller.go:250","node":"worker-us4tg-1197729831-lp2tw","state":"True","time":"17-10-26 19:54:36.734"}
{"Info":"Checking node status","caller":"kvm-operator-node-controller/controller.go:186","node":"worker-v3v91-1995990299-s6nkx","time":"17-10-26 19:54:36.734"}
{"Info":"node state","caller":"kvm-operator-node-controller/controller.go:250","node":"worker-v3v91-1995990299-s6nkx","state":"True","time":"17-10-26 19:54:36.734"}
{"Info":"Checking node status","caller":"kvm-operator-node-controller/controller.go:186","node":"master-2ir56-3922763829-nq9wf","time":"17-10-26 19:54:36.734"}
{"Info":"node state","caller":"kvm-operator-node-controller/controller.go:250","node":"master-2ir56-3922763829-nq9wf","state":"True","time":"17-10-26 19:54:36.734"}
{"Info":"Checking node status","caller":"kvm-operator-node-controller/controller.go:186","node":"worker-k9nl5-7148567-prxmm","time":"17-10-26 19:54:36.734"}
{"Info":"node state","caller":"kvm-operator-node-controller/controller.go:250","node":"worker-k9nl5-7148567-prxmm","state":"True","time":"17-10-26 19:54:36.734"}

```

File: controller.go [hunk 1]
```diff
@@ -48,6 +49,15 @@ const (
 
 	// The amount of time  between node status checks
 	nodeMonitorPeriod = 30 * time.Second
+
+	// Guest cluster environment variable names.
+	clusterAPIEnv = "GUEST_CLUSTER_API"
+	clusterIDEnv  = "GUEST_CLUSTER_ID"
```
Comment (kopiczko) [inline controller.go]: Now we have env vars and this:

```
	flag.StringVar(&kubeconfig, "kubeconfig", "", "absolute path to the kubeconfig file")
	flag.StringVar(&master, "master", "", "master url")
	flag.BoolVar(&help, "help", false, "Print usage and exit")
```

I think we should have unified way to pass config, I personally prefer flags. We can discuss that, but that's what we had so far and let's not introduce that in this PR.

File: controller.go [hunk 2]
```diff
@@ -138,13 +153,13 @@ func (c *controller) Run(stopCh chan struct{}) {
 	defer utilruntime.HandleCrash()
 
 	// Let the workers stop when we are done
-	glog.Info("Starting kvm-operator node controller")
+	c.logger.Log("Info", "Starting kvm-operator node controller")
```
Comment (kopiczko) [inline controller.go]: We have `"info"` everywhere else. We also start message with lower letter.
BTW at some point we should add `Info, Warn, Error, Debug` methods I think.

File: controller.go [hunk 3]
```diff
@@ -91,22 +102,19 @@ func main() {
 		return
 	}
 
-	// Create kubeclient config.
-	config, err := clientcmd.BuildConfigFromFlags(master, kubeconfig)
-	if err != nil {
-		glog.Fatal(err)
-	}
+	clusterAPI := os.Getenv(clusterAPIEnv)
```
Comment (JosephSalisbury) [inline controller.go]: We tend to load this kind of stuff via flags, instead of the environment

File: controller.go [hunk 4]
```diff
@@ -48,6 +49,15 @@ const (
 
 	// The amount of time  between node status checks
 	nodeMonitorPeriod = 30 * time.Second
+
+	// Guest cluster environment variable names.
+	clusterAPIEnv = "GUEST_CLUSTER_API"
+	clusterIDEnv  = "GUEST_CLUSTER_ID"
+
+	// Certificate secret keys
+	caKey  = "ca"
```
Comment (JosephSalisbury) [inline controller.go]: See https://github.com/giantswarm/certificatetpr/blob/master/tls_assets.go#L22 - please use these constants

File: controller.go (hunk 2 above)
Comment (JosephSalisbury) [inline controller.go]: we tend to use `info`, over `Info` (or `debug` is even more common)

File: controller.go [hunk 5]
```diff
@@ -234,3 +267,75 @@ func ensureNodeExists(instances cloudprovider.Instances, node *v1.Node) (bool, e
...
+	// Get API secret with certificates.
+	secretName := fmt.Sprintf("%s-api", clusterID)
+
+	secret, err := hostClient.CoreV1().Secrets(metav1.NamespaceDefault).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	// XXX(r7vme): Not using operatorkit k8s client here, because it does not support
+	// TLS certificates as raw data, only supports files.
```
Comment (kopiczko) [inline controller.go]: I think raw data should be the default supported way. We should change that. https://github.com/giantswarm/operatorkit/issues/97

File: controller.go [hunk 6]
```diff
@@ -138,13 +153,13 @@ func (c *controller) Run(stopCh chan struct{}) {
...
 	// Let the workers stop when we are done
-	glog.Info("Starting kvm-operator node controller")
+	c.logger.Log("Info", "Starting kvm-operator node controller")
 
 	// Start a loop to periodically check if any nodes have been deleted from cloudprovider
 	go wait.Until(c.MonitorNode, c.nodeMonitorPeriod, stopCh)
 
 	<-stopCh
-	glog.Info("Stopping kvm-operator node controller")
+	c.logger.Log("Info", "Stopping kvm-operator node controller")
```
Comment (JosephSalisbury) [inline controller.go]: `info`, not `Info`

File: controller.go [hunk 7]
```diff
@@ -160,15 +175,15 @@ func (c *controller) MonitorNode() {
 	// Get nodes known by kubernetes cluster.
 	nodes, err := c.kubeClient.CoreV1().Nodes().List(metav1.ListOptions{ResourceVersion: "0"})
 	if err != nil {
-		glog.Errorf("Error monitoring node status: %v", err)
+		c.logger.Log("Error", "Error monitoring node status", "Trace", microerror.Mask(err))
```
Comment (JosephSalisbury) [inline controller.go]: `error`, not `Error`

File: controller.go [hunk 8]
```diff
@@ -160,15 +175,15 @@ func (c *controller) MonitorNode() {
...
 		return
 	}
 
 	for i := range nodes.Items {
 		var currentReadyCondition *v1.NodeCondition
 		node := &nodes.Items[i]
 
-		glog.Infof("Checking node %s", node.Name)
+		c.logger.Log("Info", "Checking node status", "node", node.Name)
```
Comment (JosephSalisbury) [inline controller.go]: you get my point here :D

File: controller_test.go [hunk 9]
```diff
@@ -95,6 +96,8 @@ func TestNodeDeleted(t *testing.T) {
 		DeleteWaitChan: make(chan struct{}),
 	}
 
+	logger, _ := micrologger.New(micrologger.DefaultConfig())
```
Comment (JosephSalisbury) [inline controller_test.go]: Not a blocker, and not a big deal at all here, but I would recommend checking errors, even in tests, you can catch some weirdnesses.

File: controller.go [hunk 10]
```diff
@@ -234,3 +267,75 @@ func ensureNodeExists(instances cloudprovider.Instances, node *v1.Node) (bool, e
...
 }
+
+// If clusterAPI and clusterID are set, then we are running
+// inside Giantnetes, so try to get certificates and create
+// kubernetes client for guest cluster. If not proceed with
+// kubeconfig file.
+func newGuestClient(clusterID, clusterAPI, master, kubeconfig string) (clientset.Interface, error) {
+	if clusterAPI != "" && clusterID != "" {
+		return newGuestClientFromSecret(clusterAPI, clusterID)
+	}
```
Comment (kopiczko) [inline controller.go]: I'm unclear. Where is the controller supposed to run? In guest or host cluster? I thought only in host. If not, why is this flexible?

File: controller.go [hunk 11]
```diff
@@ -234,3 +267,75 @@ func ensureNodeExists(instances cloudprovider.Instances, node *v1.Node) (bool, e
...
+	kubeClient, err := clientset.NewForConfig(config)
+	if err != nil {
+		panic(err)
+	}
+
+	return kubeClient, nil
+}
+
+func newGuestClientFromSecret(clusterID, clusterAPI string) (clientset.Interface, error) {
+	// Get host cluster client to retrieve certificates.
```
Comment (kopiczko) [inline controller.go]: Can this https://github.com/giantswarm/certificatetpr/blob/0d95e43c5d2fba49e55308ea807ed562d358c6be/service.go#L69 be used?

**If not:**

Can we wrap it in a block? That way it is clear that we only want a secret from this block.

Something like:

```
// Get host cluster client to retrieve certificates.
var secret <package>.Secret
{

	// ...

	 secret, err = hostClient.CoreV1().Secrets(metav1.NamespaceDefault).Get(secretName, metav1.GetOptions{})
 	if err != nil {
 		return nil, microerror.Mask(err)
 	}
}

File: controller.go (hunk 10 above)
Comment (r7vme) [reply controller.go]: You're right, it should run only in host cluster. So i removed this unecessary logic. See updated change.

File: controller.go (hunk 11 above)
Comment (r7vme) [reply controller.go]: done

**Comment by r7vme** [comment]
@JosephSalisbury  @kopiczko Thanks for useful comments. Addressed all of them please take look.

**Comment by r7vme** [comment]
This version tested together with 

https://github.com/giantswarm/kvm-operator/pull/221
https://github.com/giantswarm/kubernetesd/pull/289

in `lycan`

**Review by JosephSalisbury — APPROVED** [review]
Let's roll

**Pull request merged**